### PR TITLE
Clear node collection cache after collection is done

### PR DESCRIPTION
**Pull request by nicoddemus**
Also rename the variable to convey its intent better

### CODE REVIEW

**Comment by nicoddemus** [comment]
> I think the renaming is not necessary

I find it makes the intent of the cache clearer (only meant to use during collection); unless you are against it I would like to keep it.

I will add the type hints, thanks. 👍

File: src/_pytest/main.py [hunk 1]
```diff
@@ -392,7 +393,7 @@ def __init__(self, config):
         self.startdir = config.invocation_dir
         self._initialpaths = frozenset()
         # Keep track of any collected nodes in here, so we don't duplicate fixtures
-        self._node_cache = {}
+        self._node_collection_cache = {}  # type: Dict[str, List["Node"]]
```
Comment (blueyed) [inline src/_pytest/main.py]:
```suggestion
        self._node_collection_cache = {}  # type: Dict[str, List[Node]]
```

Also `-> None` then for `__init__` if it works.

**Comment by nicoddemus** [comment]
Done, thanks for the pointers!

Had to add a few more type hints and also applied the same logic to the old `_pkg_roots` attribute. 

I want to eventually refactor the entire logic out of the `Session` class so the `session` instance is not so polluted with temporary variables/caches.

**Comment by nicoddemus** [comment]
> Is this meant to fix a correctness issue or just optimization/reduce memory usage?

The latter. 😁

**Comment by nicoddemus** [comment]
Rebased this on features too, as it should be.

**Pull request merged**